### PR TITLE
stats: support Counter and record api usage

### PR DESCRIPTION
**Pull request by yichengq**
The metrics output from cfsctl:

```
2015/06/03 05:21:51 [INFO][github.com/c-fs/cfs/cfsctl] stats_command.go:33: Metrics: [name:"cfs0_cfs_read_ops_total" val:"1"  name:"cfs0_cfs_write_ops_total" val:"1" ]
```

for #20 

It could run well both under OS X and linux.

Current metrics impl is a simple in-memory one.

### CODE REVIEW

File: stats/stats.go [hunk 1]
```diff
@@ -70,3 +51,20 @@ func (s *server) ContainerInfo(ctx context.Context, req *pb.ContainerInfoRequest
 	}
 	return &pb.ContainerInfoReply{Info: string(b)}, nil
 }
+
+func (s *server) Metrics(ctx context.Context, req *pb.MetricsRequest) (*pb.MetricsReply, error) {
+	counters, _ := metrics.Snapshot()
+	cms := make([]*pb.Metric, 0)
```
Comment (wangtuanjie) [inline stats/stats.go]: cms := make([]*pb.Metric, len(counters))

Comment (yutongp) [inline stats/stats.go]: Maybe make[(]*pb.Metric,0, len(counters))? So  you can still do append

File: cfsctl/stats_command.go [hunk 2]
```diff
@@ -24,8 +24,15 @@ var statsCmd = &cobra.Command{
 func handleStats(ctx context.Context, c pb.StatsClient) error {
 	reply, err := c.ContainerInfo(context.TODO(), &pb.ContainerInfoRequest{})
 	if err != nil || reply.Error != "" {
-		log.Fatalf("ContainerInfo err (%v)", err)
+		log.Printf("ContainerInfo err (%v)", err)
```
Comment (yutongp) [inline cfsctl/stats_command.go:25]: You may want rebase for this part

**Comment by yichengq** [comment]
rebased and addressed all comments.

File: client/client.go [hunk 3]
```diff
@@ -103,6 +103,15 @@ func (c *Client) Stats(ctx context.Context) (string, error) {
 	return reply.Info, nil
 }
 
+func (c *Client) Metrics(ctx context.Context) ([]*pb.Metric, error) {
+	reply, err := c.statsClient.Metrics(context.TODO(), &pb.MetricsRequest{})
```
Comment (yutongp) [inline client/client.go]: pass the `ctx` param?

**Comment by xiang90** [comment]
ah........ let us go back to the original question for etcd... why do you choose coda's pkg?

File: client/client.go (hunk 3 above)
Comment (yichengq) [inline client/client.go]: yes. will do.

**Comment by yichengq** [comment]
@xiang90 In this PR, i want to provide some metrics API and build the workflow of measurement. In the future, i hope this is extensible and could support disk backend and push report mode(influxdb or graphite).
So I need a simple in-memory impl for now, and coda's pkg is good enough.
I don't use prometheus because it is kind of heavy and uses pull report mode.

File: proto/stats.proto [hunk 4]
```diff
@@ -13,3 +14,15 @@ message ContainerInfoReply {
 	optional string info = 1;
 	optional string error = 2;
 }
+
+message MetricsRequest {
```
Comment (xiang90) [inline proto/stats.proto:18]: is it possible that we can pass in the name of the metrics we want? maybe we only care about cpu or io, we can say ["cpu", "io"]

File: stats/stats.go [hunk 5]
```diff
@@ -70,3 +50,20 @@ func (s *server) ContainerInfo(ctx context.Context, req *pb.ContainerInfoRequest
...
+	cms := make([]*pb.Metric, 0, len(counters))
+	for n, v := range counters {
+		cms = append(cms, &pb.Metric{Name: n, Val: strconv.FormatUint(v, 10)})
+	}
+	sort.Sort(metricsSlice(cms))
+	r := &pb.MetricsReply{Counters: cms}
+	return r, nil
+}
+
+type metricsSlice []*pb.Metric
```
Comment (xiang90) [inline stats/stats.go]: metricsByName

**Comment by xiang90** [comment]
LGTM

File: proto/stats.proto (hunk 4 above)
Comment (yichengq) [inline proto/stats.proto:18]: yes. i added it into TODO.

**Pull request merged**